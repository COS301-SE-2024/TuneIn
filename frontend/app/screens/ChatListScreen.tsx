import React, { useState } from 'react';
import { View, Text, FlatList, TouchableOpacity, Image, Dimensions, TextInput, StyleSheet } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useRouter } from 'expo-router';

const chats = [
  { id: '1', name: 'John Doe', lastMessage: 'Hey there!', avatar: 'https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1' },
  { id: '2', name: 'Jane Smith', lastMessage: 'What\'s up?', avatar: 'https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1' },
  // Add more dummy chats
];

const ChatListScreen = () => {
  const router = useRouter();
  const screenWidth = Dimensions.get('window').width;
  const [searchQuery, setSearchQuery] = useState('');

  const handleSearch = () => {
    // Handle search functionality here
    // You can filter 'chats' array based on 'searchQuery'
    // For simplicity, let's console log the searchQuery
    console.log('Searching for:', searchQuery);
  };

  return (
    
    <View style={{ flex: 1, backgroundColor: 'white', paddingHorizontal: 20, paddingTop: 20 }}>
        <View >
        <TouchableOpacity onPress={() => router.back()}>
          <Ionicons name="chevron-back" size={24} color="black" />
        </TouchableOpacity>
        </View>
      <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'center', padding: 10 }}>
        
      </View>
      <View style={styles.searchContainer}>
        <TextInput
          style={styles.searchInput}
          placeholder="Search for a user..."
          value={searchQuery}
          onChangeText={setSearchQuery}
        />
        <TouchableOpacity style={styles.searchIconContainer} onPress={handleSearch}>
          <Ionicons name="search" size={24} color="black" />
        </TouchableOpacity>
      </View>
      <FlatList
        data={chats}
        keyExtractor={item => item.id}
        renderItem={({ item }) => (
          <TouchableOpacity 
            style={{ flexDirection: 'row', alignItems: 'center', padding: 10, borderBottomWidth: 1, borderBottomColor: '#D1D5DB' }}
            onPress={() => router.push(`/screens/ChatScreen?name=${item.name}&avatar=${item.avatar}`)}
          >
            <Image source={{ uri: item.avatar }} style={{ width: 48, height: 48, borderRadius: 24, marginRight: 16 }} />
            <View style={{ flex: 1 }}>
              <Text style={{ fontSize: 16, fontWeight: 'bold' }}>{item.name}</Text>
              <Text style={{ fontSize: 14, color: 'gray' }}>{item.lastMessage}</Text>
            </View>
          </TouchableOpacity>
        )}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  searchContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    borderColor: '#CCCCCC',
    borderWidth: 1,
    borderRadius: 20,
    paddingHorizontal: 16,
    marginBottom: 20,
  },
  searchInput: {
    flex: 1,
    height: 40,
    paddingVertical: 0,
  },
  searchIconContainer: {
    padding: 10,
  },
});

export default ChatListScreen;
