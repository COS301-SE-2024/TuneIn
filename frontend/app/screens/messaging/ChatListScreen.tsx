import React, { useState, useEffect } from "react";
import {
	View,
	Text,
	FlatList,
	TextInput,
	StyleSheet,
	TouchableOpacity,
	ToastAndroid,
} from "react-native";
import { Ionicons, Entypo } from "@expo/vector-icons";
import ChatItem, { ChatItemProps } from "../../components/ChatItem";
import { colors } from "../../styles/colors";
import CreateChatScreen from "./CreateChatScreen";
import Modal from "react-native-modal";
import { useRouter } from "expo-router";
import auth from "../../services/AuthManagement";
import * as utils from "../../services/Utils";
import axios, { AxiosResponse } from "axios";
import { DirectMessageDto, UserDto, RoomDto } from "../../../api";
import { useLive } from "../../LiveContext";
import { useAPI } from "../../APIContext";

// const initialChats: Chat[] = [
// 	{
// 		id: "1",
// 		name: "John Doe",
// 		lastMessage: "Hey there!",
// 		avatar:
// 			"https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
// 	},
// 	{
// 		id: "2",
// 		name: "Jane Smith",
// 		lastMessage: "What's up?",
// 		avatar:
// 			"https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
// 	},
// 	// Add more dummy chats
// ];

const createChats = (
	messages: DirectMessageDto[],
	selfID: string,
): ChatItemProps[] => {
	const chats: ChatItemProps[] = [];
	for (const message of messages) {
		chats.push({
			message: message,
			otherUser:
				message.sender.userID === selfID ? message.recipient : message.sender,
		});
	}
	return chats;
};

const ChatListScreen = () => {
	const { currentUser, refreshUser, setRefreshUser } = useLive();
	const { rooms } = useAPI();
	const [searchQuery, setSearchQuery] = useState("");
	const [userMessages, setUserMessages] = useState<DirectMessageDto[]>([]);
	const [messageRooms, setMessageRooms] = useState<(RoomDto | undefined)[]>([]);
	const [filteredChats, setFilteredChats] = useState<ChatItemProps[]>([]);
	const [friends, setFriends] = useState<UserDto[]>([]);
	const [isModalVisible, setModalVisible] = useState(false);
	const router = useRouter();

	useEffect(() => {
		// Fetch chats from backend
		(async () => {
			if (!currentUser) {
				if (!refreshUser) {
					setRefreshUser(true);
				}
				return;
			}
			try {
				const token = await auth.getToken();

				let promises = [];
				promises.push(
					axios.get(`${utils.API_BASE_URL}/users/dms`, {
						headers: { Authorization: `Bearer ${token}` },
					}),
				);
				promises.push(
					axios.get(`${utils.API_BASE_URL}/users/friends`, {
						headers: { Authorization: `Bearer ${token}` },
					}),
				);

				const responses = await Promise.all(promises);
				const chats = responses[0].data as DirectMessageDto[];
				console.log(chats);
				setFriends(responses[1].data as UserDto[]);
				console.log(responses[1].data);

				const roomPromises: (
					| Promise<AxiosResponse<RoomDto, any>>
					| Promise<undefined>
				)[] = [];
				for (const message of chats) {
					if (message.bodyIsRoomID) {
						//roomIDs.push(message.messageBody);
						roomPromises.push(rooms.getRoomInfo(message.messageBody));
					} else {
						roomPromises.push(Promise.resolve(undefined));
					}
				}
				const roomResponses = await Promise.all(roomPromises);
				const tmpRooms: (RoomDto | undefined)[] = [];
				for (const roomResponse of roomResponses) {
					if (roomResponse !== undefined) {
						tmpRooms.push(roomResponse.data as RoomDto);
					} else {
						tmpRooms.push(undefined);
					}
				}
				setMessageRooms(tmpRooms);
				setFilteredChats(createChats(chats, currentUser.userID));
				setUserMessages(chats);
			} catch (error) {
				console.log(error);
				ToastAndroid.show("Failed to load DMs", ToastAndroid.SHORT);
				throw error;
			}
		})();
	}, [currentUser, rooms, refreshUser, setRefreshUser]);

	useEffect(() => {
		/*
		if (searchQuery === "") {
			setFilteredChats(initialChats);
		} else {
			const filtered = initialChats.filter((chat) =>
				chat.name.toLowerCase().includes(searchQuery.toLowerCase()),
			);
			setFilteredChats(filtered);
		}
			*/
		if (searchQuery === "") {
			if (currentUser !== undefined) {
				setFilteredChats(createChats(userMessages, currentUser.userID));
			}
		} else {
			if (currentUser !== undefined) {
				const filtered = userMessages.filter((chat) => {
					return chat.sender.profile_name
						.toLowerCase()
						.includes(searchQuery.toLowerCase());
				});

				setFilteredChats(createChats(filtered, currentUser.userID));
			}
		}
	}, [currentUser, searchQuery, userMessages]);

	const toggleModal = () => {
		setModalVisible(!isModalVisible);
	};

	return (
		<View style={styles.screenContainer}>
			<View style={styles.headerContainer}>
				<TouchableOpacity testID="back-button" onPress={() => router.back()}>
					<Ionicons name="chevron-back" size={24} color="black" />
				</TouchableOpacity>
				<Text style={styles.chatHeader}>Chats</Text>
			</View>
			<View style={styles.searchContainer}>
				<TextInput
					style={styles.searchInput}
					placeholder="Search for a user..."
					value={searchQuery}
					onChangeText={setSearchQuery}
				/>
				<TouchableOpacity style={styles.searchIconContainer} onPress={() => {}}>
					<Ionicons name="search" size={24} color="black" />
				</TouchableOpacity>
			</View>
			<FlatList
				data={filteredChats}
				keyExtractor={(item) => item.message.pID}
				renderItem={({ item }) => (
					<ChatItem message={item.message} otherUser={item.otherUser} />
				)}
			/>
			<TouchableOpacity style={styles.newChatButton} onPress={toggleModal}>
				<Entypo name="message" size={24} color="white" />
			</TouchableOpacity>
			<Modal
				isVisible={isModalVisible}
				onBackdropPress={toggleModal}
				onSwipeComplete={toggleModal}
				swipeDirection="down"
				style={styles.modal}
			>
				<CreateChatScreen closeModal={toggleModal} friends={friends} />
			</Modal>
		</View>
	);
};

const styles = StyleSheet.create({
	screenContainer: {
		flex: 1,
		backgroundColor: "white",
		paddingHorizontal: 20,
		paddingTop: 20,
	},
	headerContainer: {
		flexDirection: "row",
		alignItems: "center",
		padding: 10,
	},
	chatHeader: {
		flex: 1,
		fontSize: 24,
		fontWeight: "bold",
		textAlign: "center",
	},
	searchContainer: {
		flexDirection: "row",
		alignItems: "center",
		borderColor: "#CCCCCC",
		borderWidth: 1,
		borderRadius: 20,
		paddingHorizontal: 16,
		marginBottom: 20,
		marginTop: 10,
	},
	searchInput: {
		flex: 1,
		height: 40,
		paddingVertical: 0,
	},
	searchIconContainer: {
		padding: 10,
	},
	newChatButton: {
		position: "absolute",
		right: 20,
		bottom: 20,
		width: 60,
		height: 60,
		borderRadius: 30,
		backgroundColor: colors.primary,
		alignItems: "center",
		justifyContent: "center",
	},
	modal: {
		justifyContent: "flex-end",
		margin: 0,
		height: "90%",
	},
});

export default ChatListScreen;
