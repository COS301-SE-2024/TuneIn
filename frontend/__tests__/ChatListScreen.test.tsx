import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import ChatListScreen from "../app/screens/messaging/ChatListScreen";
import { useRouter } from "expo-router";
import { TextInput, TouchableOpacity } from "react-native";

jest.mock("expo-font", () => ({
	...jest.requireActual("expo-font"),
	loadAsync: jest.fn(),
}));

jest.mock("expo-asset", () => ({
	...jest.requireActual("expo-asset"),
	fromModule: jest.fn(() => ({
		downloadAsync: jest.fn(),
		uri: "mock-uri",
	})),
}));

jest.mock("expo-router", () => ({
	useRouter: jest.fn(),
}));

describe("ChatListScreen", () => {
	const mockRouter = {
		push: jest.fn(),
		back: jest.fn(),
	};

	beforeEach(() => {
		(useRouter as jest.Mock).mockReturnValue(mockRouter);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it("renders the chat list screen correctly", () => {
		const { toJSON } = render(<ChatListScreen />);
		expect(toJSON()).toMatchSnapshot();
	});

	it("handles search input correctly", () => {
		const { getByPlaceholderText } = render(<ChatListScreen />);
		const searchInput = getByPlaceholderText("Search for a user..."); // Adjust placeholder to match your component

		fireEvent.changeText(searchInput, "John");

		expect(searchInput.props.value).toBe("John");
	});

	it("navigates back when back button is pressed", () => {
		const { getAllByTestId } = render(<ChatListScreen />);
		const backButton = getAllByTestId("back-button")[0]; // Assuming the back button has a testID of 'back-button'

		fireEvent.press(backButton);

		expect(mockRouter.back).toHaveBeenCalled();
	});
});
