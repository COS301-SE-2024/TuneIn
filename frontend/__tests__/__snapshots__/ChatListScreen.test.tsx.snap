// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatListScreen renders the chat list screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 10,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="back-button"
    >
      <Text />
    </View>
    <Text
      style={
        {
          "flex": 1,
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Chats
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#CCCCCC",
        "borderRadius": 20,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 20,
        "marginTop": 10,
        "paddingHorizontal": 16,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search for a user..."
      style={
        {
          "flex": 1,
          "height": 40,
          "paddingVertical": 0,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text />
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "avatar": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
          "id": "1",
          "lastMessage": "Hey there!",
          "name": "John Doe",
        },
        {
          "avatar": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
          "id": "2",
          "lastMessage": "What's up?",
          "name": "Jane Smith",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#D1D5DB",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "margin": 10,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
              }
            }
            style={
              {
                "borderRadius": 24,
                "height": 48,
                "marginRight": 16,
                "width": 48,
              }
            }
          />
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              John Doe
            </Text>
            <Text
              style={
                {
                  "color": "gray",
                  "fontSize": 14,
                }
              }
            >
              Hey there!
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#D1D5DB",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "margin": 10,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
              }
            }
            style={
              {
                "borderRadius": 24,
                "height": 48,
                "marginRight": 16,
                "width": 48,
              }
            }
          />
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              Jane Smith
            </Text>
            <Text
              style={
                {
                  "color": "gray",
                  "fontSize": 14,
                }
              }
            >
              What's up?
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#08BDBD",
        "borderRadius": 30,
        "bottom": 20,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 20,
        "width": 60,
      }
    }
  >
    <Text />
  </View>
</View>
`;
